Make `rosbag2_transport::Player::play()` run in a separate thread

roncapat opened the pull request:
- Closes #1500 

This is intended to solve the issue reported in #1500.
CC: @MichaelOrlov @KenYN

This can be also beneficial introductory work for #1419. As it has been noticed, playback shall start automatically when the `Player` component is loaded, but this can't be done via syncronous call by the component constructor.

Current status:
- separate thread implemented
- all tests preliminarly adapted

Anyway, I believe that a round of refactor can be done here. For example, direclty returning an `std::future` from `wait_for_playback_to_end()`. Opinions?

@KenYN may I ask you also to do a test with this branch to confirm that the `play` service now returns immediately?

roncapat committed:
Play spawns a separate thread

Signed-off-by: Patrick Roncagliolo <ronca.pat@gmail.com>

KenYN commented:
I did `git checkout patch-2 && git merge roncagliolo/async-play` and it appears to be operating correctly, although by no means is this a thorough test, but thank you very much for this.

morlov-apexai committed:
Address multiple race conditions in player and some renames after review

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Fixes in the rosbag2_transport tests

- Fixed multiple race conditions in tests when
wait_for_playback_to_end() calls in a separate thread before
Player::play() method. Those race conditions would likely lead to a
false negative or flaky tests.

- Replaced player_future with wait_for_playback_to_end() to the direct
call of the player_->wait_for_playback_to_end() in the most of the
places.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Rename wait_for_playback_to_end() to the wait_for_playback_to_finish()

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Use rcpputils::unique_lock for RCPPUTILS TSA guarded mutexes

- Rationale: To support Clang Thread Safety Analysis
- Removed locally defined TSAUniqueLock which is a duplicate of the
rcpputils::unique_lock.
- Also removed some unused includes in player.cpp and player.hpp

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add timeout parameter to the wait_for_playback_to_finish(timeout)

- Rationale: Make wait_for_playback_to_finish(..) to be optionally as
a non-blocking call

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov reviewed (changes requested):
@roncapat See my suggestions for review fixes in the https://github.com/roncapat/rosbag2/pull/6 PR

roncapat commented:
@MichaelOrlov review branch merged :)

MichaelOrlov reviewed (approved):
LGTM

MichaelOrlov commented:
Gist: https://gist.githubusercontent.com/MichaelOrlov/aa32bdbe242c8c06daca683655e76bb4/raw/4be47e99df1fdad56dd18ce5f2d2c4f04ff13980/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_transport rosbag2_tests rosbag2_py
TEST args:  --packages-above rosbag2_transport rosbag2_tests rosbag2_py
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/12985
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=19896)](http://ci.ros2.org/job/ci_linux/19896/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=14380)](http://ci.ros2.org/job/ci_linux-aarch64/14380/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=20603)](http://ci.ros2.org/job/ci_windows/20603/)

MichaelOrlov commented:
https://github.com/Mergifyio backport iron

MichaelOrlov commented:
Note:
Backporting to the Iron and Humble is not possible because It will cause ABI-breaking changes.
-  it requires adding new member variables to the Player class and we don't have `pimpl` design for the Player class on the Iron and Humble branches.
